includes/admin/upgrades/v3/class-data-migrator.php

```diff
@@ -455,35 +455,67 @@ public static function orders( $data = null ) {
 			return;
 		}
 
-		$subtotal = 0;
-		$tax      = 0;
-		$discount = 0;
-		$total    = 0;
+		$subtotal    = 0;
+		$order_tax   = 0;
+		$discount    = 0;
+		$order_total = 0;
+
+		// Track the total value of added fees in case the Order was initially migrated
+		// without _edd_payment_total or _edd_payment_tax and manual calculation was needed.
+		$order_fees_tax         = 0;
+		$order_fees_total       = 0;
+		$order_items_fees_tax   = 0;
+		$order_items_fees_total = 0;
+
+		// Retrieve the tax amount from metadata if available.
+		$meta_tax = isset( $meta['_edd_payment_tax'] )
+			? $meta['_edd_payment_tax']
+			: false;
+
+		if ( false !== $meta_tax ) {
+			$meta_tax  = maybe_unserialize( $meta_tax );
+			$order_tax = (float) $meta_tax[0];
+		}
+
+		// Retrieve the total amount from metadata if available.
+		$meta_total = isset( $meta['_edd_payment_total'] )
+			? $meta['_edd_payment_total']
+			: false;
+
+		if ( false !== $meta_total ) {
+			$meta_total  = maybe_unserialize( $meta_total );
+			$order_total = (float) $meta_total[0];
+		}
 
 		// In some cases (very few) there is no cart details...so we have to just avoid this part.
 		if ( ! empty( $cart_details ) && is_array( $cart_details ) ) {
 
 			// Loop through the items in the purchase to build the totals.
 			foreach ( $cart_details as $cart_item ) {
-				$subtotal += isset( $cart_item['subtotal'] ) ? (float) $cart_item['subtotal'] : 0;
-				$tax      += isset( $cart_item['tax'] )      ? (float) $cart_item['tax']      : 0;
 				$discount += isset( $cart_item['discount'] ) ? (float) $cart_item['discount'] : 0;
-				$total    += isset( $cart_item['price'] )    ? (float) $cart_item['price']    : 0;
-			}
+				$subtotal += isset( $cart_item['subtotal'] ) ? (float) $cart_item['subtotal'] : 0;
 
-		} else {
+				// Add the cart line item tax amount if a total is not available on the order.
+				if ( false === $meta_tax ) {
+					$order_tax += isset( $cart_item['tax'] ) ? (float) $cart_item['tax'] : 0;
+				}
 
-			// As a backup, we can get some information from other meta keys.
-			if ( isset( $meta['_edd_payment_total'][0] ) ) {
-				$total = (float) $meta['_edd_payment_total'][0];
+				// Add the cart line item price amount (includes tax, order item fee, _but not order item fee tax_)
+				// if a total is not available on the order.
+				if ( false === $meta_total ) {
+					$order_total += isset( $cart_item['price'] ) ? (float) $cart_item['price'] : 0;
+				}
 			}
 
-			if ( isset( $meta['_edd_payment_tax'][0] ) ) {
-				$tax = (float) $meta['_edd_payment_tax'][0];
-			}
+			// Cart details cannot be retrieved.
+		} else {
 
-			$subtotal = $total - $tax;
+			// We lose the discount amount at this point... it remains 0.
 
+			// Calculate the subtotal if we have a total and tax amount.
+			if ( false !== $meta_tax && false !== $meta_total ) {
```

Comment:
I'm not sure this should be in the `else` (which would mean no `else` is required. In the orders I'm checking for a similar issue with discounts, this PR is correctly setting the order total, but not the subtotal, which is `0`. There are cart details but they're incomplete, so the order(s) are not getting to this point.

Reply:
@robincornett So the cart details exist, but don't contain a `subtotal` entry for each item? I believe `release/3.0` would still result in a 0 subtotal?

Maybe we could check if subtotal is still `0` after the cart loop and subtract any found tax and discounts from the total and use that?

Reply:
I've logged it in issue #8643 ... it's somewhat related to what you're doing here, and that PR does fix the subtotal issue (mainly by making sure that the cart details are normalized before using, as they're used twice, but potentially adjusted before using in `release/3.0`).

Here's the cart details from the affected order (I do have more):

```
[cart_details] => a:1:{i:0;a:5:{s:4:"name";s:29:"PayPal Pro and PayPal Express";s:2:"id";s:3:"162";s:11:"item_number";a:2:{s:2:"id";s:3:"162";s:7:"options";a:0:{}}s:5:"price";s:5:"49.00";s:8:"quantity";i:1;}}
```

I think we need to fix both issues separately, but your PR should be finalized first and merged and then I can update mine, because I also was tinkering with order totals and your approach is better.